File: learnSphere/frontend/learnsphere/src/app/core/services/discussion.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from '@core/services/api.service';
import { DiscussionPostDto, DiscussionPostRequest } from '@core/models/social.model';

/**
 * DiscussionService — shared across all surfaces that render course/lecture
 * discussion threads: the student/teacher course-wide discussion pages and
 * the per-lecture discussion widget embedded in the lecture player.
 */
@Injectable({
  providedIn: 'root'
})
export class DiscussionService {
  constructor(private api: ApiService) {}

  getCourseDiscussion(courseId: number): Observable<DiscussionPostDto[]> {
    return this.api.get<DiscussionPostDto[]>(`/courses/${courseId}/discussion`);
  }

  getLectureDiscussion(lectureId: number): Observable<DiscussionPostDto[]> {
    return this.api.get<DiscussionPostDto[]>(`/lectures/${lectureId}/discussion`);
  }

  createPost(payload: DiscussionPostRequest): Observable<DiscussionPostDto> {
    return this.api.post<DiscussionPostDto>('/discussion', payload);
  }

  deletePost(id: number): Observable<void> {
    return this.api.delete<void>(`/discussion/${id}`);
  }
}

File: learnsphere-backend/src/main/java/com/MOOC/OnlineLearningPlatfrom/Service/DiscussionService.java
package com.MOOC.OnlineLearningPlatfrom.Service;

import com.MOOC.OnlineLearningPlatfrom.Dto.DiscussionPostDto;
import com.MOOC.OnlineLearningPlatfrom.Dto.DiscussionPostRequestDto;
import com.MOOC.OnlineLearningPlatfrom.Security.CustomUserDetails;

import java.util.List;

public interface DiscussionService {
    List<DiscussionPostDto> getCourseDiscussion(Long courseId);
    List<DiscussionPostDto> getLectureDiscussion(Long lectureId);
    DiscussionPostDto createPost(DiscussionPostRequestDto request, CustomUserDetails principal);
    void deletePost(Long id, CustomUserDetails principal);
}
